### PR TITLE
Feat/appl records/v45

### PR DESCRIPTION
**Pull request by victorjulien**
#6669 where `AppLayerStream` is now `StreamSlice` to better reflect what it is. Removes previous use of thread local vars to handle the records, in favor of using the `StreamSlice` data.

suricata-verify-pr: 594

### CODE REVIEW

File: src/app-layer-frames.c [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+
+    Frame *r = FrameNew(frames, (uint32_t)ptr_offset, len);
+    if (r != NULL) {
+        r->type = frame_type;
+    }
+    return r;
+}
+
+Frame *AppLayerFrameNew2(Flow *f, const StreamSlice *stream_slice, const uint32_t frame_start_rel,
+        const uint32_t len, int dir, uint8_t frame_type)
```
Comment (jasonish) [inline src/app-layer-frames.c:350]: Will there just be one variation when done?

File: src/app-layer-frames.c [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+    /* workarounds for many (unit|fuzz)tests not handling TCP data properly */
+#if defined(UNITTESTS) || defined(FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION)
+    if (f->protoctx == NULL)
+        return NULL;
+    if (frame_start < stream_slice->input || frame_start >= stream_slice->input + stream_slice->input_len)
+        return NULL;
+#endif
+    BUG_ON(frame_start < stream_slice->input);
+    BUG_ON(stream_slice->input == NULL);
+    BUG_ON(f->proto != IPPROTO_TCP);
```
Comment (jasonish) [inline src/app-layer-frames.c:319]: No UDP support? I got tripped up while looking at DNS frame support which can be TCP or UDP.

**Comment by jasonish** [comment]
Looking at the logging for a crude DNS conversion I did, I see this logged:
```
"frame": {
    "type": "message",
    "id": 0,
    "offset": 0,
    "length": 34,
    "diframetion": "toserver",
    "complete": true,
    "payload": "ACAAAAEAAAEAAAAAAAADd3d3Bmdvb2dsZQNjb20AAAEAAQ==",
    "payload_printable": ". .............www.google.com.....",
    "payload_hex": "|00 20 00 00 01 00 00 01 00 00 00 00 00 00 03|www|06|google|03|com|00 00 01 00 01|"
  }
```

Out side of debug, is the `id` needed? Same with `offset`, and to a lesser extend `length`.

Also, interesting field, `diframetion`? Would the direction be needed with that other work that Eric has been doing that tells us more about the direction of an alert?

File: src/detect-engine.c [hunk 3]
```diff
@@ -920,6 +920,31 @@ const char *DetectBufferTypeGetDescriptionByName(const char *name)
...
+    BUG_ON(!exists);
+    exists->mpm = true;
+    SCLogDebug("%p %s -- %d supports mpm", exists, name, exists->id);
+}
+
+void DetectEngineBufferTypeSupportsTransformations(DetectEngineCtx *de_ctx, const char *name)
+{
+    DetectBufferType *exists = DetectEngineBufferTypeLookupByName(de_ctx, name);
+    BUG_ON(!exists);
+    BUG_ON(!exists);
```
Comment (inashivb) [inline src/detect-engine.c:1234]: nit: Extra `BUG_ON`?

File: src/app-layer-frames.c (hunk 2 above)
Comment (victorjulien) [reply src/app-layer-frames.c:319]: > Looking at the logging for a crude DNS conversion I did, I see this logged:
> ...
> Out side of debug, is the `id` needed? Same with `offset`, and to a lesser extend `length`.

Offset I think tells you something useful: its the start of the frame vs the start of the stream. Id I'm not sure about whether its useful.

> Also, interesting field, `diframetion`? Would the direction be needed with that other work that Eric has been doing that tells us more about the direction of an alert?

That should be `direction`. Its a artifact of `s/rec/frame/g` ;)

**Comment by victorjulien** [comment]
> Looking at the logging for a crude DNS conversion I did, I see this logged:
> ...
> Out side of debug, is the `id` needed? Same with `offset`, and to a lesser extend `length`.

Offset I think tells you something useful: its the start of the frame vs the start of the stream. Id I'm not sure about whether its useful.

> Also, interesting field, `diframetion`? Would the direction be needed with that other work that Eric has been doing that tells us more about the direction of an alert?

That should be `direction`. Its a artifact of `s/rec/frame/g` ;)

File: src/app-layer-frames.c (hunk 2 above)
Comment (victorjulien) [reply src/app-layer-frames.c:319]: I've focused on TCP for now, as this is where the perf gain is to be had, so blocked everything else. But I don't think its hard to extend this to UDP. The data the frames point to will be in the actual packet payload then.

File: src/app-layer-frames.c (hunk 1 above)
Comment (victorjulien) [reply src/app-layer-frames.c:350]: Hope so

**Comment by victorjulien** [comment]
`length` I think is useful. Think about trying to detect an issue there a record has a specific length and you want to use `frame:dns.message; bsize:666;`. Think it helps if suri just tells the rule writer/analyst what the len is.

File: src/app-layer-parser.c [hunk 4]
```diff
@@ -1219,11 +1234,10 @@ int AppLayerParserParse(ThreadVars *tv, AppLayerParserThreadCtx *alp_tctx, Flow
 
     /* invoke the recursive parser, but only on data. We may get empty msgs on EOF */
     if (input_len > 0 || (flags & STREAM_EOF)) {
+        Setup(f, flags & (STREAM_TOSERVER | STREAM_TOCLIENT), input, input_len, &stream_slice);
```
Comment (inashivb) [inline src/app-layer-parser.c:1285]: Q: So, if we are at the end of a stream, we might end up creating an empty stream slice here?

File: src/app-layer-frames.c [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+            // fall through
+        } else {
+            if (frames->dyn_size == 256) {
+                SCLogNotice("limit reached! 256 dynamic frames already");
+                // limit reached
+                return NULL;
+            }
+
+            /* realloc time */
+            uint16_t new_dyn_size = frames->dyn_size * 2;
```
Comment (inashivb) [inline src/app-layer-frames.c:117]: Q: Can it become > 256 at this stage? Say, we already have 129 dynamic frames. Is that permissible?

File: src/app-layer-frames.c [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+        return &frames->dframes[dyn_cnt];
+    }
+}
+
+static void FrameClean(Frame *frame)
+{
+    memset(frame, 0, sizeof(*frame));
+}
+
+static void StreamDPUCopy(Frame *dst, Frame *src)
```
Comment (inashivb) [inline src/app-layer-frames.c:144]: Q: Is this a typo or does it really mean something? `DPU`

File: src/app-layer-frames.c [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+            if (frame->rel_offset + frame->len <= (int32_t)slide) {
+                // remove by not incrementing 'x'
+                SCLogDebug("removing %p", frame);
+                FrameClean(frame);
+            } else {
+                Frame *nframe;
+                if (x >= FRAMES_STATIC_CNT) {
+                    nframe = &frames->dframes[x - FRAMES_STATIC_CNT];
+                } else {
+                    nframe = &frames->sframes[x];
```
Comment (inashivb) [inline src/app-layer-frames.c:232]: Q: Why do we set a static frame here if we are in the dynamic frame zone?

File: src/app-layer-frames.h [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+#include "detect-engine-state.h"
+#include "util-file.h"
+#include "stream-tcp-private.h"
+#include "rust.h"
+#include "app-layer-parser.h"
+
+typedef int64_t FrameId;
+
+enum {
+    FRAME_FLAGE_ENDS_AT_EOF,
```
Comment (inashivb) [inline src/app-layer-frames.h:37]: nit: typo `FLAGE`?

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 5147

File: src/app-layer-frames.c (hunk 1 above)
Comment (victorjulien) [reply src/app-layer-frames.c:350]: Kept 2 in #6684, but gave them clearer names

File: src/app-layer-parser.c (hunk 4 above)
Comment (victorjulien) [reply src/app-layer-parser.c:1285]: Yes but that is fine. The parser should be accepting that as an EOF notification.

File: src/app-layer-frames.h (hunk 8 above)
Comment (victorjulien) [reply src/app-layer-frames.h:37]: its deliberate. The FLAGE values are set to 0-N by their position in the enum, then the define the next line turns them into proper flags. Open to suggestions on the naming.

File: src/app-layer-frames.c (hunk 5 above)
Comment (victorjulien) [reply src/app-layer-frames.c:117]: the check above limits it to 256, which is arbitrary at this point. It starts at 8, and only doubles. So I don't see how we could exceed 256 here

File: src/app-layer-frames.c (hunk 6 above)
Comment (victorjulien) [reply src/app-layer-frames.c:144]: left over from older naming, will fix

File: src/app-layer-frames.c (hunk 7 above)
Comment (victorjulien) [reply src/app-layer-frames.c:232]: we're moving this after removing frames from our list, so we can move frames from dynamic to static

**Comment by victorjulien** [comment]
Replaced by #6684. Some comments will be addressed later as I only saw some of the comments now.

**Pull request closed without merging**